akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,443 @@
+    lastHandshakeStatus = result.getHandshakeStatus
+    if (tracing) log.debug(s"wrap: status=${result.getStatus} handshake=$lastHandshakeStatus remaining=${userInBuffer.remaining} out=${transportOutBuffer.position}")
+    if (lastHandshakeStatus == FINISHED) handshakeFinished()
+    runDelegatedTasks()
+    result.getStatus match {
+      case OK ⇒
+        flushToTransport()
+        userInChoppingBlock.putBack(userInBuffer)
+      case CLOSED ⇒
+        flushToTransport()
+        if (engine.isInboundDone()) nextPhase(completedPhase)
+        else nextPhase(awaitingClose)
+      case s ⇒ fail(new IllegalStateException(s"unexpected status $s in doWrap()"))
+    }
+  }
+
+  @tailrec
+  private def doUnwrap(ignoreOutput: Boolean = false): Unit = {
+    val result = engine.unwrap(transportInBuffer, userOutBuffer)
+    if (ignoreOutput) userOutBuffer.clear()
+    lastHandshakeStatus = result.getHandshakeStatus
+    if (tracing) log.debug(s"unwrap: status=${result.getStatus} handshake=$lastHandshakeStatus remaining=${transportInBuffer.remaining} out=${userOutBuffer.position}")
+    runDelegatedTasks()
+    result.getStatus match {
+      case OK ⇒
+        result.getHandshakeStatus match {
+          case NEED_WRAP ⇒ flushToUser()
+          case FINISHED ⇒
+            flushToUser()
+            handshakeFinished()
+            transportInChoppingBlock.putBack(transportInBuffer)
+          case _ ⇒
+            if (transportInBuffer.hasRemaining()) doUnwrap()
+            else flushToUser()
+        }
+      case CLOSED ⇒
+        flushToUser()
+        if (engine.isOutboundDone()) nextPhase(completedPhase)
+        else nextPhase(flushingOutbound)
+      case BUFFER_UNDERFLOW ⇒
+        flushToUser()
+      case BUFFER_OVERFLOW ⇒
+        flushToUser()
+        transportInChoppingBlock.putBack(transportInBuffer)
+      case s ⇒ fail(new IllegalStateException(s"unexpected status $s in doUnwrap()"))
+    }
+  }
+
+  @tailrec
+  private def runDelegatedTasks(): Unit = {
+    val task = engine.getDelegatedTask
+    if (task != null) {
+      if (tracing) log.debug("running task")
+      task.run()
+      runDelegatedTasks()
+    } else {
+      val st = lastHandshakeStatus
+      lastHandshakeStatus = engine.getHandshakeStatus
+      if (tracing && st != lastHandshakeStatus) log.debug(s"handshake status after tasks: $lastHandshakeStatus")
+    }
+  }
+
+  private def handshakeFinished(): Unit = {
+    if (tracing) log.debug("handshake finished")
+    currentSession = engine.getSession
+    corkUser = false
+  }
+
+  override def receive = LoggingReceive(inputBunch.subreceive.orElse[Any, Unit](outputBunch.subreceive))
+
+  nextPhase(bidirectional)
+
+  protected def fail(e: Throwable): Unit = {
+    // FIXME: escalate to supervisor
+    if (tracing) log.debug("fail {} due to: {}", self, e.getMessage)
+    inputBunch.cancel()
+    outputBunch.error(TransportOut, e)
+    outputBunch.error(UserOut, e)
+    context.stop(self)
```

Comment:
No, indeed I did not, thanks for the reminder! My thinking was that this stage can only fail once everything is up and running, but now I see that this is not true: if we get onError before everything is up and running then we terminate early—good catch! Since you fixed those: if this happens TLS is out of the picture and it boils down to letting the pump handle it, right? In this case, could you take care of this one as well? Otherwise I’ll pick it up later when I’m finished with the japi stuff.